File: app/src/main/java/com/android/ashwiask/tvmaze/shows/ShowsPagedAdaptor.java
package com.android.ashwiask.tvmaze.shows;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.android.ashwiask.tvmaze.R;
import com.android.ashwiask.tvmaze.databinding.LoadingListItemBinding;
import com.android.ashwiask.tvmaze.databinding.NetworkFailureListItemBinding;
import com.android.ashwiask.tvmaze.databinding.ShowListItemBinding;
import com.android.ashwiask.tvmaze.home.Show;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.bumptech.glide.request.RequestOptions;

public class ShowsPagedAdaptor extends PagedListAdapter<Show, RecyclerView.ViewHolder> {
    private Context context;
    private NetworkState currentNetworkState;

    ShowsPagedAdaptor(@NonNull DiffUtil.ItemCallback<Show> diffCallback) {
        super(diffCallback);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        context = parent.getContext();
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        switch (viewType) {
            case R.layout.loading_list_item:
                LoadingListItemBinding loadingListItemBinding =
                        LoadingListItemBinding.inflate(layoutInflater, parent, false);
                return new LoadingHolder(loadingListItemBinding);
            case R.layout.network_failure_list_item:
                NetworkFailureListItemBinding networkFailureListItemBinding =
                        NetworkFailureListItemBinding.inflate(layoutInflater, parent, false);
                return new NetworkFailureHolder(networkFailureListItemBinding);
            case R.layout.show_list_item:
            default:
                ShowListItemBinding showListItemBinding =
                        ShowListItemBinding.inflate(layoutInflater, parent, false);
                return new ShowHolder(showListItemBinding);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        switch (getItemViewType(position)) {
            case R.layout.loading_list_item:
                LoadingHolder loadingHolder = (LoadingHolder) holder;
                loadingHolder.binding.spinner.setVisibility(View.VISIBLE);
                break;
            case R.layout.network_failure_list_item:
                NetworkFailureHolder networkHolder = (NetworkFailureHolder) holder;
                networkHolder.binding.networkPbm.setText(currentNetworkState.message());
                break;
            default:
            case R.layout.show_list_item:
                Show show = getItem(position);
                configureImage((ShowHolder) holder, show);
                break;
        }
    }

    private void configureImage(ShowHolder holder, @NonNull Show show) {
        if (show.image() != null) {
            Glide.with(context).load(show.image().get("original"))
                    .apply(RequestOptions.placeholderOf(R.color.grey))
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(holder.binding.showImage);
        }
    }

    public void setNetworkState(NetworkState newNetworkState) {
        NetworkState previousNetworkState = currentNetworkState;
        boolean hadExtraRow = hasExtraRow();
        currentNetworkState = newNetworkState;
        boolean hasExtraRow = hasExtraRow();
        if (hadExtraRow != hasExtraRow) {
            if (hadExtraRow) {
                notifyItemRemoved(super.getItemCount());
            } else {
                notifyItemInserted(super.getItemCount());
            }
        } else if (hasExtraRow && !previousNetworkState.equals(newNetworkState)) {
            notifyItemChanged(getItemCount() - 1);
        }
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + getExtraRow();
    }

    // Add an item for the network state
    private int getExtraRow() {
        if (hasExtraRow()) return 1;
        else return 0;
    }

    @Override
    public int getItemViewType(int position) {
        // Reached at the end
        if (hasExtraRow() && position == getItemCount() - 1) {
            if (currentNetworkState.status() == NetworkState.Status.LOADING) {
                return R.layout.loading_list_item;
            } else {
                return R.layout.network_failure_list_item;
            }
        } else {
            return R.layout.show_list_item;
        }
    }

    private boolean hasExtraRow() {
        return currentNetworkState != null && currentNetworkState.status() != NetworkState.Status.SUCCESS;
    }

    static class ShowHolder extends RecyclerView.ViewHolder {
        private ShowListItemBinding binding;

        ShowHolder(ShowListItemBinding itemBinding) {
            super(itemBinding.getRoot());
            this.binding = itemBinding;
        }
    }

    static class LoadingHolder extends RecyclerView.ViewHolder {
        private LoadingListItemBinding binding;

        LoadingHolder(LoadingListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    static class NetworkFailureHolder extends RecyclerView.ViewHolder {
        private NetworkFailureListItemBinding binding;

        NetworkFailureHolder(NetworkFailureListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/android/ashwiask/tvmaze/shows/ShowsPagedAdaptor.kt
package com.android.ashwiask.tvmaze.shows

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.android.ashwiask.tvmaze.R
import com.android.ashwiask.tvmaze.databinding.AllShowListItemBinding
import com.android.ashwiask.tvmaze.databinding.LoadingListItemBinding
import com.android.ashwiask.tvmaze.databinding.NetworkFailureListItemBinding
import com.android.ashwiask.tvmaze.network.home.Show
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions

class ShowsPagedAdaptor constructor(
    diffCallback: DiffUtil.ItemCallback<Show>,
    private val callback: Callback
) : PagedListAdapter<Show, RecyclerView.ViewHolder>(diffCallback) {
    private lateinit var context: Context
    private var currentNetworkState: NetworkState = Loading

    // Add an item for the network state
    private val extraRow: Int
        get() = if (hasExtraRow())
            1
        else
            0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        context = parent.context
        val layoutInflater = LayoutInflater.from(context)
        when (viewType) {
            R.layout.loading_list_item -> {
                val loadingListItemBinding = LoadingListItemBinding.inflate(layoutInflater, parent, false)
                return LoadingHolder(loadingListItemBinding)
            }
            R.layout.network_failure_list_item -> {
                val networkFailureListItemBinding = NetworkFailureListItemBinding.inflate(layoutInflater, parent, false)
                val networkFailureHolder = NetworkFailureHolder(networkFailureListItemBinding)
                networkFailureHolder.binding.retry.setOnClickListener { callback.onRetryClicked() }
                return networkFailureHolder
            }
            else -> {
                val showListItemBinding = AllShowListItemBinding.inflate(layoutInflater, parent, false)
                return ShowHolder(showListItemBinding)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            R.layout.loading_list_item -> {
                val loadingHolder = holder as LoadingHolder
                loadingHolder.binding.spinner.visibility = View.VISIBLE
            }
            R.layout.network_failure_list_item -> {
                val networkHolder = holder as NetworkFailureHolder
                val networkError = currentNetworkState as NetworkError
                networkHolder.binding.networkPbm.text = networkError.message
            }
            R.layout.show_list_item -> {
                val show = getItem(position)
                val showHolder = holder as ShowHolder
                showHolder.binding.show = show
                if (show!!.rating != null) {
                    showHolder.binding.rating = show.rating!!["average"]
                }
                configureImage(showHolder, show)
            }
        }
    }

    private fun configureImage(holder: ShowHolder, show: Show) {
        if (show.image != null) {
            Glide.with(context).load(show.image["original"])
                .apply(RequestOptions.placeholderOf(R.color.grey))
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(holder.binding.showImage)
        }
    }

    fun setNetworkState(newNetworkState: NetworkState) {
        val previousNetworkState = currentNetworkState
        val hadExtraRow = hasExtraRow()
        currentNetworkState = newNetworkState
        val hasExtraRow = hasExtraRow()
        if (hadExtraRow != hasExtraRow) {
            if (hadExtraRow) {
                notifyItemRemoved(super.getItemCount())
            } else {
                notifyItemInserted(super.getItemCount())
            }
        } else if (hasExtraRow && previousNetworkState != newNetworkState) {
            notifyItemChanged(itemCount - 1)
        }
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + extraRow
    }

    override fun getItemViewType(position: Int): Int {
        // Reached at the end
        return if (hasExtraRow() && position == itemCount - 1) {
            if (currentNetworkState === Loading) {
                R.layout.loading_list_item
            } else {
                R.layout.network_failure_list_item
            }
        } else {
            R.layout.show_list_item
        }
    }

    private fun hasExtraRow(): Boolean {
        return currentNetworkState == Success
    }

    class ShowHolder(val binding: AllShowListItemBinding) : RecyclerView.ViewHolder(binding.root)

    class LoadingHolder(val binding: LoadingListItemBinding) : RecyclerView.ViewHolder(binding.root)

    class NetworkFailureHolder(val binding: NetworkFailureListItemBinding) :
        RecyclerView.ViewHolder(binding.root)

    interface Callback {
        fun onRetryClicked()
    }
}
